### PR TITLE
Multiple keymanager support

### PR DESCRIPTION
**Pull request by tharindu1st**
resolves #https://github.com/wso2/product-apim/issues/8209

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java [hunk 1]
```diff
@@ -307,13 +308,31 @@
...
+    INVALID_TENANT(901300,"Tenant Not Found", 400, "Tenant Not Found"),
+    // Key Manager Related
+    INVALID_KEY_MANAGER_TYPE(901400, "Key Manager Type not configured", 400, "Key Manager Type not configured"),
+    REQUIRED_KEY_MANAGER_CONFIGURATION_MISSING(901401,"Required Key Manager configuration missing",400,"Missing " +
+            "required configuration"),
+    KEY_MANAGER_ALREADY_EXIST(901402, "Key Manager Already Exist", 409, "Key Manager Already Exist"),
+    KEY_MANAGER_NOT_FOUND(901403, "Key Manager Not found", 400, "Key Manager not found"), KEY_MANAGER_NAME_EMPTY(901404,
+            "Key Manager name cannot be empty", 400,
+            "Key Manager name cannot be empty"),
+    KEY_MANAGER_NOT_SUPPORT_OAUTH_APP_CREATION(901404, "Key Manager Didn't support to Generate Client Application", 400,
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java:320]: Make all messages present tense

Comment (bhathiya) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java:320]: @shaniR can you please review all the messages?

Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java:320]:
```suggestion
    KEY_MANAGER_NOT_SUPPORT_OAUTH_APP_CREATION(901404, "Key Manager doesn't support OAuth application generation", 400,
```

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java [hunk 2]
```diff
@@ -307,13 +308,31 @@
...
+            "Key Manager Didn't support to Generate Client Application"),
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java:321]: same here

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java [hunk 3]
```diff
@@ -307,13 +308,31 @@
...
+    KEY_MANAGER_NOT_SUPPORTED_TOKEN_GENERATION(901405, "Key Manager didn't support to generate Token ", 400,
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java:322]:
```suggestion
    KEY_MANAGER_NOT_SUPPORTED_TOKEN_GENERATION(901405, "Key manager doesn't support token generation", 400,
```

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java [hunk 4]
```diff
@@ -307,13 +308,31 @@
...
+            "Key Manager didn't support to generate Token "),
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java:323]: same here

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java [hunk 5]
```diff
@@ -307,13 +308,31 @@
...
+    KEY_MANAGER_NOT_ENABLED(901406, "Key Manager not enabled in the system", 400,
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java:324]:
```suggestion
    KEY_MANAGER_NOT_ENABLED(901406, "Key Manager is not enabled in the system", 400,
```

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java [hunk 6]
```diff
@@ -307,13 +308,31 @@
...
+            "Key Manager not enabled in the system"),
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java:325]: same here

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java [hunk 7]
```diff
@@ -307,13 +308,31 @@
...
+    KEY_MANAGER_MISSING_REQUIRED_PROPERTIES_IN_APPLICATION(901407, "Required Application Properties missing", 400,
+            "Required Application Properties missing"),
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java:327]:
```suggestion
    KEY_MANAGER_MISSING_REQUIRED_PROPERTIES_IN_APPLICATION(901407, "Required application properties are missing", 400,
            "Required application properties are missing"),
```

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java [hunk 8]
```diff
@@ -307,13 +308,31 @@
...
+    KEY_MAPPING_ALREADY_EXIST(901408, "Application Already Registered", 409, "Application Already Registered");
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java:328]:
```suggestion
    KEY_MAPPING_ALREADY_EXIST(901408, "Oauth application is already created", 409, "Oauth application is already created");
```

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/JWTUtil.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+            throw new APIManagementException("Error retrieving certificate from truststore ",e);
+        }
+
+        if (publicCert != null) {
+            JWSAlgorithm algorithm = jwt.getHeader().getAlgorithm();
+            if ((JWSAlgorithm.RS256.equals(algorithm) || JWSAlgorithm.RS512.equals(algorithm) ||
+                    JWSAlgorithm.RS384.equals(algorithm))) {
+                return verifyTokenSignature(jwt, (RSAPublicKey) publicCert.getPublicKey());
+            } else {
+                log.error("Public key is not a RSA");
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/JWTUtil.java:95]:
```suggestion
                log.error("Public key is not RSA");
```

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/JWTUtil.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+                throw new APIManagementException("Public key is not a RSA");
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/JWTUtil.java:96]:
```suggestion
                throw new APIManagementException("Public key is not RSA");
```

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/JWTUtil.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+            }
+        } else {
+            log.error("Couldn't find a public certificate to verify signature with alias " + alias);
+            throw new APIManagementException("Couldn't find a public certificate to verify signature with alias ");
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/JWTUtil.java:100]:
```suggestion
            log.error("Couldn't find a public certificate with alias " + alias + " to verify the signature");
            throw new APIManagementException("Couldn't find a public certificate with alias " + alias + " to verify the signature");
```

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java [hunk 12]
```diff
@@ -742,6 +737,12 @@ public static API getAPIForPublishing(GovernanceArtifact artifact, Registry regi
             api.setApiCategories(getAPICategoriesFromAPIGovernanceArtifact(artifact, tenantId));
             //get endpoint config string from artifact, parse it as a json and set the environment list configured with
             //non empty URLs to API object
+            String keyManagers = artifact.getAttribute(APIConstants.API_OVERVIEW_KEY_MANAGERS);
+            if (StringUtils.isNotEmpty(keyManagers)){
+                api.setKeyManagers(new Gson().fromJson(keyManagers,List.class));
+            }else{
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java:743]: format the entire code

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/KeyManager.java [hunk 13]
```diff
@@ -318,42 +314,41 @@ default void detachResourceScopes(API api, Set<URITemplate> uriTemplates, String
...
      */
-    default void validateScopes(Set<Scope> scopes, String tenantDomain) throws APIManagementException {
+    default void validateScopes(Set<Scope> scopes) throws APIManagementException {
         // Doing nothing in default implementation. If KM supports validate scopes operation, override the
         // implementation.
     }
 
+    public String getType();
+
+    public void setTenantDomain(String tenantDomain);
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/KeyManager.java:353]: add java docs

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/KeyManagerConnectorConfiguration.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+package org.wso2.carbon.apimgt.api.model;
+
+import java.util.List;
+
+public interface KeyManagerConnectorConfiguration {
+    public String getImplementation();
+    public String getJWTValidator();
+    public List<ConfigurationDto> getConnectionConfigurations();
+    public List<ConfigurationDto> getApplicationConfigurations();
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/KeyManagerConnectorConfiguration.java:27]: java docs.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java [hunk 15]
```diff
@@ -1084,6 +1082,7 @@ private CORSHeaders() {
         public static final String CLIENT_DETAILS = "appDetails";
         public static final String CALLBACK_URL = "callbackUrl";
         public static final String KEY_STATE = "keyState";
+        public static final String KEY_MAPPING_UD = "keyMappingId";
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java:1085]:
```suggestion
        public static final String KEY_MAPPING_ID = "keyMappingId";
```

**Comment by isharac** [comment]
Shall we add a design doc/ flow chat as a reference?

**Pull request merged**